tensorflow_quantum/python/optimizers/rotosolve_minimizer.py

```diff
@@ -0,0 +1,284 @@
+    ])
+
+
+def _get_initial_state(initial_position, tolerance, expectation_value_function):
+    """Create RotosolveOptimizerResults with initial state of search ."""
+    init_args = {
+        "converged": tf.Variable(False),
+        "num_iterations": tf.Variable(0),
+        "num_objective_evaluations": expectation_value_function.iter,
+        "position": tf.Variable(initial_position),
+        "objective_value": tf.Variable(0.),
+        "last_objective_value": tf.Variable(0.),
+        "tolerance": tolerance,
+        "solve_param_i": tf.Variable(0)
+    }
+    return RotosolveOptimizerResults(**init_args)
+
+
+def minimize(expectation_value_function,
+             initial_position,
+             tolerance=1e-8,
+             max_iterations=50,
+             name=None):
+    """Applies the rotosolve algorithm to minimize a linear combination
+     of quantum measurement expectation values. See arXiv:1903.12166,
+      arXiv:1905.09692
+
+    ### Usage:
+
+    The following example demonstrates the Rotosolve optimizer attempting
+     to find the minimum for two qubit ansatz expectation value.
+
+    ```python
+
+    # In this example we train a circuit perform an XOR operation
+    X = np.asarray([
+        [0, 0],
+        [0, 1],
+        [1, 0],
+        [1, 1],
+    ], dtype=float)
+
+    Y = np.asarray([
+        [-1], [1], [1], [-1]
+    ], dtype=float)
+
+    def convert_to_circuit(input_data):
+        # Encode into quantum datapoint.
+        values = np.ndarray.flatten(input_data)
+        qubits = cirq.GridQubit.rect(1, 2)
+        circuit = cirq.Circuit()
+        for i, value in enumerate(values):
+            if value:
+                circuit.append(cirq.X(qubits[i]))
+        return circuit
+
+    x_circ = tfq.convert_to_tensor([convert_to_circuit(x) for x in X])
+
+    # Create two qubits
+    q0, q1 = cirq.GridQubit.rect(1, 2)
+
+    # Create an anzatz on these qubits.
+    a, b = sympy.symbols('a b') # parameters for the circuit
+    circuit = cirq.Circuit(
+        cirq.rx(a).on(q0),
+        cirq.ry(b).on(q1), cirq.CNOT(control=q0, target=q1))
+
+    # Build the Keras model.
+    model = tf.keras.Sequential([
+        # The input is the data-circuit, encoded as a tf.string
+        tf.keras.layers.Input(shape=(), dtype=tf.string),
+        # The PQC layer returns the expected value of the
+        # readout gate, range [-1,1].
+        tfq.layers.PQC(circuit, cirq.Z(q1)),
+    ])
+
+    def hinge_loss(y_true, y_pred):
+        # Here we use hinge loss as the cost function
+        tf.reduce_mean(tf.cast(1 - y_true * y_pred, tf.float32))
+
+    result = rotosolve_minimizer.minimize(
+        rotosolve_minimizer.function_factory(
+            model, hinge_loss, x_circ, Y), np.random.rand([2])
+            ) # Initial guess of the parameter
+
+    print(result)
+    ```
+
+    Args:
+      expectation_value_function:  A Python callable that accepts
+        a point as a real `Tensor` and returns a tuple of `Tensor`s

```

Comment:
Nit: `tf.Tensor`. Can you also outline the shape requirements for the input as well as the returned shape you get from the output ?

Reply:
Fixed in f2b15d449ea0ddd1870b903ee914a92a001e5dc2